gazenot/src/error.rs

```diff
@@ -0,0 +1,70 @@
+use miette::Diagnostic;
+use thiserror::Error;
+
+pub type Result<T> = std::result::Result<T, GazenotError>;
+pub type ResultInner<T> = std::result::Result<T, GazenotErrorInner>;
+
+#[derive(Error, Debug, Diagnostic)]
+#[error("couldn't {operation}")]
+pub struct GazenotError {
+    pub operation: String,
+    #[help]
+    pub help: Option<String>,
+    #[source]
+    #[diagnostic_source]
+    pub cause: GazenotErrorInner,
+}
+
+impl GazenotError {
+    pub fn new(operation: impl Into<String>, err: impl Into<GazenotErrorInner>) -> Self {
+        Self {
+            operation: operation.into(),
+            help: None,
+            cause: err.into(),
+        }
+    }
+    pub fn with_url(
+        operation: impl Into<String>,
+        url: impl std::fmt::Display,
+        err: impl Into<GazenotErrorInner>,
+    ) -> Self {
+        Self {
+            operation: operation.into(),
+            help: Some(format!("was accessing this endpoint: {url}")),
+            cause: err.into(),
+        }
+    }
+}
+
+#[derive(Error, Debug, Diagnostic)]
+#[error("{0}")]
+pub struct SimpleError(pub String);
+
+#[derive(Error, Debug, Diagnostic)]
+pub enum GazenotErrorInner {
+    #[error(transparent)]
+    Reqwest(#[from] reqwest::Error),
+    #[error(transparent)]
+    HeaderValue(#[from] reqwest::header::InvalidHeaderValue),
+    #[error(transparent)]
+    Join(#[from] tokio::task::JoinError),
+    #[error(transparent)]
+    Url(#[from] url::ParseError),
+    #[error(transparent)]
+    Asset(#[from] axoasset::AxoassetError),
+    #[error("server error {status}")]
+    ResponseError {
+        status: reqwest::StatusCode,
+        #[related]
+        errors: Vec<SimpleError>,
```

Comment:
This is just wrapping strings from the server?

Reply:
Yeah, the schema allows the server to return a list of errors. miette/thiserror isn't willing to say "a string is a type of error" so you need to newtype them.